multimodal/health_multimodal/image/model/model.py

```diff
@@ -0,0 +1,167 @@
+            projected_global_embedding = torch.mean(projected_patch_embeddings, dim=(2, 3))
+
+        logits = self.classifier(pooled_x) if self.classifier else None
+        return ImageModelOutput(img_embedding=pooled_x,
+                                patch_embedding=patch_x,
+                                class_logits=logits,
+                                projected_patch_embeddings=projected_patch_embeddings,
+                                projected_global_embedding=projected_global_embedding)
+
+    def create_downstream_classifier(self, **kwargs: Any) -> ClassifierHeadType:
+        """
+        Creates the classification module for the downstream task
+        """
+        downstream_classifier_kwargs = kwargs if kwargs else self.downstream_classifier_kwargs
+        return MTModel(self.feature_size, **downstream_classifier_kwargs)
+
+    @torch.no_grad()
+    def get_patchwise_projected_embeddings(self, input_img: torch.Tensor, normalize: bool) -> torch.Tensor:
+        """
+        Get patchwise projected embeddings from the CNN model.
+        Args:
+            input_img: input tensor image [B, C, H, W]
+            normalize: If set to True, the embeddings are l2-normalized.
+        Returns:
+            projected_embeddings: tensor of embeddings in shape [batch, n_patches_h, n_patches_w, feature_size]
+        """
+        assert self.training is False, "This function is only implemented for evaluation mode"
+        outputs = self.forward(input_img)
+        projected_embeddings = outputs.projected_patch_embeddings.detach()  # type: ignore
+        if normalize:
+            projected_embeddings = F.normalize(projected_embeddings, dim=1)
+        projected_embeddings = projected_embeddings.permute([0, 2, 3, 1])  # B D H W -> B H W D (D: Features)
+        return projected_embeddings
+
+
+class ImageEncoder(nn.Module):
+    def __init__(self, img_model_type: str):
+        super().__init__()
+        self.img_model_type = img_model_type
+        self.encoder = self._create_encoder()
+
+    def _create_encoder(self, **kwargs: Any) -> nn.Module:
+        if self.img_model_type == 'resnet18':
+            encoder = resnet18(return_all_feature_maps=True, pretrained=True, **kwargs)
+        elif self.img_model_type == 'resnet50':
+            encoder = resnet50(return_all_feature_maps=True, pretrained=True, **kwargs)
+        else:
+            raise NotImplementedError
+
+        return encoder
+
+    def forward(self, x: torch.Tensor, return_patch_embeddings: bool = False) -> Union[T, Tuple[T, T]]:
+        x = self.encoder(x)
+        x = x[-1] if isinstance(x, list) else x
+        avg_pooled_emb = torch.flatten(torch.nn.functional.adaptive_avg_pool2d(x, (1, 1)), 1)
+        if return_patch_embeddings:
+            return x, avg_pooled_emb
+        else:
+            return avg_pooled_emb
+
+    def reload_encoder_with_dilation(self, replace_stride_with_dilation: List[bool] = [False, False, True]) -> None:
```

Comment:
Since we allow the model to be in training mode, I think it is worth adding a warning that this should be called _before_ initializing the optimizer otherwise this will cause a disconnect between the params in the model and the params in the optimizer.